eli5/keras/gradcam.py

```diff
@@ -0,0 +1,206 @@
+        Keras layer instance to differentiate with respect to.
+    
+
+    See :func:`eli5.keras.explain_prediction` for description of the 
+    ``estimator``, ``doc``, ``targets`` parameters.
+
+    Returns
+    -------
+    (weights, activations, gradients, predicted_idx, predicted_val) : (numpy.ndarray, ..., int, float)
+        Values of variables.
+    """
+    # score for class in targets
+    predicted_idx = _get_target_prediction(targets, estimator)
+    predicted_val = K.gather(estimator.output[0,:], predicted_idx) # access value by index
+
+    # output of target activation layer, i.e. activation maps of a convolutional layer
+    activation_output = activation_layer.output 
+
+    # score for class w.r.p.t. activation layer
+    grads = _calc_gradient(predicted_val, [activation_output])
+
+    # Global Average Pooling of gradients to get the weights
+    # note that axes are in range [-rank(x), rank(x)) (we start from 1, not 0)
+    # TODO: decide whether this should go in gradcam_backend() or gradcam()
+    weights = K.mean(grads, axis=(1, 2))
+
+    evaluate = K.function([estimator.input], 
+        [weights, activation_output, grads, predicted_val, predicted_idx]
+    )
+    # evaluate the graph / do actual computations
+    weights, activations, grads, predicted_val, predicted_idx = evaluate([doc])
+    
+    # put into suitable form
+    weights = weights[0]
+    predicted_val = predicted_val[0]
+    predicted_idx = predicted_idx[0]
+    activations = activations[0, ...]
+    grads = grads[0, ...]
+    return weights, activations, grads, predicted_idx, predicted_val
+
+
+def _calc_gradient(ys, xs):
+    # (K.variable, list) -> K.variable
+    """
+    Return the gradient of scalar ``ys`` with respect to each of list ``xs``,
+    (must be singleton)
+    and apply grad normalization.
+    """
+    # differentiate ys (scalar) with respect to each variable in xs
+    grads = K.gradients(ys, xs)
+
+    # grads gives a python list with a tensor (containing the derivatives) for each xs
+    # to use grads with other operations and with K.function
+    # we need to work with the actual tensors and not the python list
+    grads, = grads # grads should be a singleton list (because xs is a singleton)
+
+    # validate that the gradients were calculated successfully (no None's)
+    # https://github.com/jacobgil/keras-grad-cam/issues/17#issuecomment-423057265
+    # https://github.com/tensorflow/tensorflow/issues/783#issuecomment-175824168
+    if grads is None:
+        raise ValueError('Gradient calculation resulted in None values. '
+                         'Check that the model is differentiate and try again. '
+                         'ys: {}. xs: {}. grads: {}'.format(
+                            ys, xs, grads))
+
+    # this seems to make the heatmap less noisy
+    grads =  K.l2_normalize(grads) 
+    return grads
+
+
+def _get_target_prediction(targets, estimator):
+    # type: (Union[None, list], Model) -> K.variable
```

Comment:
Optional[List]

Reply:
Ah I was thinking whether to use this. My thinking was that `targets` is not actually optional in the parameter list so it might be confusing. But I see that the rest of the library uses Optional (i.e. for functions that could return None) so I will change it.

https://github.com/TeamHG-Memex/eli5/pull/315/commits/e8a34f10874bf7ae563566cd6d82e69a229aa144